# Coverage Analysis - Sudipta

> COMP 4710 Group 11 - Progress Report 1

## Objective

- Compute coverage statistics across neighbourhoods
- Identify underserved areas (bottom quartile)
- Detect coverage outliers using IQR method
- Compare coverage across community areas

In [20]:
# automatically relodes modules when they are changed
# no need to restart kernel when coverage files are changed
%load_ext autoreload
%autoreload 2


The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [21]:
# Standard imports
import pandas as pd
import matplotlib.pyplot as plt

# Project imports
from ptn_analysis import get_stops_per_neighbourhood, get_stops_per_community

#import the functions from ptn_analysis.analysis.coverage
#1.compute_coverage_statistics,
#2.identify_underserved_neighbourhoods,
#3.detect_coverage_outliers,
#4.compute_coverage_by_area,
#5.compare_community_coverage,
import ptn_analysis.analysis.coverage as coverage


print("Imports loaded successfully")

Imports loaded successfully


In [22]:
# Load data
neighbourhood_data = get_stops_per_neighbourhood()
community_data = get_stops_per_community()
print(f"Loaded {len(neighbourhood_data)} neighbourhoods and {len(community_data)} communities")

Loaded 237 neighbourhoods and 12 communities


In [23]:
# Preview data
neighbourhood_data.head()

neighbourhood  area_km2  stop_count  stops_per_km2
0               Rossmere-A  3.978106          64      16.088058
1     St. James Industrial  4.646884          61      13.127076
2             Windsor Park  3.353475          60      17.891888
3         River Park South  4.421495          59      13.343903
4  Inkster Industrial Park  2.936009          57      19.414107

## Coverage Statistics

In [24]:
# Compute summary statistics
stats = coverage.compute_coverage_statistics()
print("Coverage Statistics:")
for key, value in stats.items():
    print(f"  {key}: {value}")

Coverage Statistics:
  total_neighbourhoods: 237
  total_stops: 3873
  mean_stops: 16
  median_stops: 12
  std_stops: 14
  min_stops: 0
  max_stops: 64
  zero_stop_areas: 16
  mean_density_per_km2: 12.800988850370716


## Underserved Neighbourhoods

In [25]:
# Find underserved areas (bottom 25%)
underserved = coverage.identify_underserved_neighbourhoods(threshold_percentile=25.0)
print(f"Found {len(underserved)} underserved neighbourhoods:")
underserved

Found 60 underserved neighbourhoods:


neighbourhood   area_km2  stop_count  stops_per_km2
225              Maple Grove Park   1.018090           0       0.000000
226                       Griffin   0.605693           0       0.000000
227     St. Vital Perimeter South  36.815616           0       0.000000
228                  Wilkes South  23.366101           0       0.000000
229                Cloutier Drive   0.820553           0       0.000000
230                 Kil-Cona Park   2.864876           0       0.000000
231            Saskatchewan North  13.738621           0       0.000000
232                      Perrault   1.961152           0       0.000000
233                Turnbull Drive   1.096329           0       0.000000
234               Armstrong Point   0.265781           0       0.000000
235                   La Barriere  14.876205           0       0.000000
236          West Perimeter South   4.801803           0       0.000000
224              Assiniboia Downs   3.910790           0       0.000000
223                    Kensington   0.105249           0       0.000000
222                    Trappistes   6.417752           0       0.000000
221         North Transcona Yards   0.871373           0       0.000000
217           Rosser-Old Kildonan  15.983132           1       0.062566
216               Symington Yards   3.706974           1       0.269762
219               Transcona North   1.905987           1       0.524663
186              Assiniboine Park   4.818367           5       1.037696
197               Ridgewood South   3.055066           4       1.309301
220               River West Park   0.763362           1       1.309995
165               Transcona South   5.697757           8       1.404061
215                      The Mint   0.694438           1       1.440013
218                     Woodhaven   0.600958           1       1.664011
78                        Airport  10.339471          18       1.740901
184              Bridgwater Lakes   1.941665           5       2.575109
214                   Niakwa Park   0.750583           2       2.664595
210             Kingston Crescent   0.717650           2       2.786872
204                  Normand Park   1.042213           3       2.878490
155                Southland Park   2.740683           8       2.918981
209                 Kildonan Park   0.966443           3       3.104165
178               Waverley West B   1.907733           6       3.145095
79   St. Boniface Industrial Park   5.540396          18       3.248865
171                     Kirkfield   1.984569           7       3.527215
163      West Kildonan Industrial   2.229307           8       3.588559
110               Transcona Yards   3.405734          13       3.817092
192                     Tyne-Tees   1.282863           5       3.897533
44              Tuxedo Industrial   6.625917          26       3.923985
120      North Inkster Industrial   2.744367          12       4.372593
199                      Archwood   0.911062           4       4.390481
212           Dufferin Industrial   0.441654           2       4.528434
203             Springfield South   0.644079           3       4.657811
157                      Glenwood   1.626813           8       4.917591
126                        Tuxedo   2.354579          12       5.096453
66                     Sage Creek   4.084806          21       5.141003
189           Wellington Crescent   0.971148           5       5.148546
149           North River Heights   1.724490           9       5.218935
164                        Dugald   1.531247           8       5.224501
196             Victoria Crescent   0.763239           4       5.240823
194                     Jameswood   0.749726           4       5.335284
150                   Stock Yards   1.650451           9       5.453055
133                      Eric Coy   2.000063          11       5.499828
206                       Lavalee   0.531845           3       5.640744
168                    Seven Oaks   1.225409           7    

## Outlier Detection

In [26]:
# Detect outliers using IQR method
outliers = coverage.detect_coverage_outliers(method='iqr')
print(f"Found {len(outliers)} outliers:")
outliers

Found 15 outliers:


neighbourhood  stop_count outlier_type
0                Rossmere-A          64         High
1      St. James Industrial          61         High
2              Windsor Park          60         High
3          River Park South          59         High
4   Inkster Industrial Park          57         High
5              Linden Woods          56         High
6           Dakota Crossing          55         High
7             Fort Richmond          54         High
8              Tyndall Park          53         High
9                  Chalmers          51         High
10                Southdale          49         High
11               River East          48         High
12                Jefferson          48         High
13          Canterbury Park          47         High
14               The Maples          47         High

## Coverage by Area

In [27]:
# Rank neighbourhoods by density
coverage_ranked = coverage.compute_coverage_by_area()
print("Top 10 by density:")
coverage_ranked.head(10)

Top 10 by density:


neighbourhood  area_km2  stop_count  stops_per_km2  density_rank
127     Portage-Ellice  0.169350          12      70.858992             1
120       Central Park  0.232539          12      51.604280             2
176     Portage & Main  0.119114           6      50.372064             3
55              Spence  0.483418          22      45.509228             4
137           Edgeland  0.270360          11      40.686446             5
165       Civic Centre  0.198625           8      40.276997             6
125  Exchange District  0.319155          12      37.599312             7
102   St. Vital Centre  0.429758          15      34.903398             8
19      West Alexander  1.395339          43      30.816889             9
214         China Town  0.065942           2      30.329668            10

In [28]:
print("Bottom 10 by density:")
coverage_ranked.tail(10)

Bottom 10 by density:


neighbourhood   area_km2  stop_count  stops_per_km2  \
227  St. Vital Perimeter South  36.815616           0            0.0   
228             Turnbull Drive   1.096329           0            0.0   
229       West Perimeter South   4.801803           0            0.0   
230            Armstrong Point   0.265781           0            0.0   
231           Assiniboia Downs   3.910790           0            0.0   
232               Wilkes South  23.366101           0            0.0   
233                   Perrault   1.961152           0            0.0   
234      North Transcona Yards   0.871373           0            0.0   
235                La Barriere  14.876205           0            0.0   
236           Maple Grove Park   1.018090           0            0.0   

     density_rank  
227           222  
228           222  
229           222  
230           222  
231           222  
232           222  
233           222  
234           222  
235           222  
236           222

## Community Comparison

In [29]:
# Compare communities
community_stats = coverage.compare_community_coverage()
community_stats

community   area_km2  stop_count  stops_per_km2  \
3                 Downtown  16.311507         364      22.315535   
11           Point Douglas  10.864817         197      18.131920   
8                  Inkster  18.083426         245      13.548318   
9            River Heights  18.088211         232      12.826033   
1               River East  34.652921         404      11.658469   
7                Transcona  28.754517         265       9.215943   
2             St. Boniface  47.616673         389       8.169407   
0               Fort Garry  76.990095         628       8.156893   
6               Seven Oaks  39.845054         300       7.529165   
4   St. James - Assiniboia  59.221982         340       5.741111   
5                St. Vital  63.231456         301       4.760289   
10       Assiniboine South  61.191992         208       3.399138   

    density_rank coverage_category  
3              1              High  
11             2              High  
8              3              High  
9              4              High  
1              5              High  
7              6              High  
2              7              High  
0              8              High  
6              9              High  
4             10              High  
5             11            Medium  
10            12            Medium

## Summary

In [30]:
stats = coverage.compute_coverage_statistics()
underserved = coverage.identify_underserved_neighbourhoods(threshold_percentile=25.0)

summary = {
    "Total neighbourhoods": stats["total_neighbourhoods"],
    "Mean stops/neighbourhood": round(stats["mean_stops"], 2),
    "Zero-stop areas": stats["zero_stop_areas"],
    "Underserved count (bottom 25%)": len(underserved),
}

summary_df = pd.DataFrame(
    summary.items(),
    columns=["Metric", "Value"]
)

summary_df



Metric  Value
0            Total neighbourhoods    237
1        Mean stops/neighbourhood     16
2                 Zero-stop areas     16
3  Underserved count (bottom 25%)     60

In [31]:
# Top 10 / Bottom 10 by stop_count (for report)
top10_stops = neighbourhood_data.sort_values("stop_count", ascending=False).head(10)
bottom10_stops = neighbourhood_data.sort_values("stop_count", ascending=True).head(10)

print("Top 10 neighbourhoods by stop_count:")
top10_stops[["neighbourhood", "stop_count", "area_km2", "stops_per_km2"]]

Top 10 neighbourhoods by stop_count:


neighbourhood  stop_count  area_km2  stops_per_km2
0               Rossmere-A          64  3.978106      16.088058
1     St. James Industrial          61  4.646884      13.127076
2             Windsor Park          60  3.353475      17.891888
3         River Park South          59  4.421495      13.343903
4  Inkster Industrial Park          57  2.936009      19.414107
5             Linden Woods          56  3.609936      15.512742
6          Dakota Crossing          55  3.832735      14.350066
7            Fort Richmond          54  4.599052      11.741551
8             Tyndall Park          53  3.013996      17.584630
9                 Chalmers          51  2.520368      20.235144

In [32]:
print("\nBottom 10 neighbourhoods by stop_count:")
bottom10_stops[["neighbourhood", "stop_count", "area_km2", "stops_per_km2"]]


Bottom 10 neighbourhoods by stop_count:


neighbourhood  stop_count   area_km2  stops_per_km2
221         Cloutier Drive           0   0.820553            0.0
222          Kil-Cona Park           0   2.864876            0.0
223             Trappistes           0   6.417752            0.0
236   West Perimeter South           0   4.801803            0.0
235            La Barriere           0  14.876205            0.0
234       Maple Grove Park           0   1.018090            0.0
233                Griffin           0   0.605693            0.0
232         Turnbull Drive           0   1.096329            0.0
231        Armstrong Point           0   0.265781            0.0
230  North Transcona Yards           0   0.871373            0.0

## Export Results

Save results to reports/figures/

In [33]:
# Example: Save underserved to CSV
from pathlib import Path

out_dir = Path("../reports/figures/coverage_reports")
out_dir.mkdir(parents=True, exist_ok=True)

top10_stops.to_csv("../reports/figures/coverage_reports/top10_neighbourhoods_by_stops.csv", index=False)
bottom10_stops.to_csv("../reports/figures/coverage_reports/bottom10_neighbourhoods_by_stops.csv", index=False)
# underserved.to_csv("reports/figures/underserved_neighbourhoods.csv", index=False)

In [34]:
# Recompute key tables (so exports are up to date)
underserved = coverage.identify_underserved_neighbourhoods(threshold_percentile=25.0)
outliers = coverage.detect_coverage_outliers(method="iqr")
community_stats = coverage.compare_community_coverage()

# Export for report / sharing
underserved.to_csv("../reports/figures/coverage_reports/underserved_neighbourhoods_bottom25.csv", index=False)
outliers.to_csv("../reports/figures/coverage_reports/coverage_outliers_iqr.csv", index=False)
community_stats.to_csv("../reports/figures/coverage_reports/community_coverage_comparison.csv", index=False)

print("Exported:")
print(" - reports/figures/top10_neighbourhoods_by_stops.csv.csv")
print(" - reports/figures/bottom10_neighbourhoods_by_stops.csv.csv")
print(" - reports/figures/underserved_neighbourhoods_bottom25.csv")
print(" - reports/figures/coverage_outliers_iqr.csv")
print(" - reports/figures/community_coverage_comparison.csv")


Exported:
 - reports/figures/top10_neighbourhoods_by_stops.csv.csv
 - reports/figures/bottom10_neighbourhoods_by_stops.csv.csv
 - reports/figures/underserved_neighbourhoods_bottom25.csv
 - reports/figures/coverage_outliers_iqr.csv
 - reports/figures/community_coverage_comparison.csv


In [35]:
# Histogram of neighbourhood stop density (stops per km^2)
# Histogram of neighbourhood stop density (stops per km^2)
plt.figure(figsize=(10, 6))

plt.hist(
    df["stops_per_km2"].dropna(),
    bins=30,
    edgecolor="black",
    alpha=0.8
)

plt.title("Distribution of Neighbourhood Stop Density", fontsize=14, pad=12)
plt.xlabel("Stops per km²", fontsize=12)
plt.ylabel("Number of neighbourhoods", fontsize=12)

plt.grid(axis="y", linestyle="--", alpha=0.6)
plt.tight_layout()

# Save for report (SAVE BEFORE SHOW)
plt.savefig("../reports/figures/coverage_reports/neighbourhood_density_hist.png", dpi=200)

plt.show()

print("Saved plot to reports/figures/neighbourhood_density_hist.png")



NameError: name 'df' is not defined

<Figure size 1000x600 with 0 Axes>

# Methodology Note

Neighbourhood/community areas were approximated from polygon area in degrees² using latitude-based km conversion (Winnipeg ~50°N) to compute stops per km².

## Interpretation

Stop coverage is uneven across Winnipeg neighbourhoods. While the average is about 16 stops per neighbourhood (237 total neighbourhoods), coverage is concentrated in certain areas and absent in others. For example, the highest stop-count neighbourhoods include Rossmere-A (64 stops), St. James Industrial (61), Windsor Park (60), River Park South (59), and Inkster Industrial Park (57), indicating strong stop presence in those zones. In contrast, several neighbourhoods have zero stops (e.g., Cloutier Drive, Assiniboia Downs, Kil-Cona Park, Perrault), which aligns with the broader result that there are 16 zero-stop neighbourhoods and 60 underserved neighbourhoods (bottom quartile by stop density). This suggests transit stop access is highly variable by location and that a meaningful subset of neighbourhoods may be underserved.

#### 1. Non-trivial number of zero-service areas

16 neighbourhoods have 0 stops, which strongly indicates areas with no stop coverage (or coverage outside the polygon boundaries). These should be flagged as priority candidates for “underserved” discussion and for checking whether they are parks/industrial/perimeter zones.

#### 2. Underserved areas are common (bottom quartile)

The bottom-quartile analysis identifies 60 neighbourhoods as underserved by stop density. Many of these are near-zero density, suggesting limited walk-up access to stops in those areas compared to the city median.

#### 3. Coverage is skewed; a few neighbourhoods dominate stop counts

Outlier detection found 15 high outlier neighbourhoods by stop count (IQR method). This indicates stop placement is concentrated in a small set of neighbourhoods (likely major corridors, transfer points, and dense residential/commercial areas), while most neighbourhoods sit far below those levels.

---
**Results by : Sudipta Sarker (7910584) COMP 4710 - University of Manitoba**